src/properties.ts

```diff
@@ -0,0 +1,347 @@
+ */
+function getBootstrapperProperties(startTimestampMs: number): ScannerProperties {
+  return {
+    'sonar.scanner.app': SCANNER_BOOTSTRAPPER_NAME,
+    'sonar.scanner.appVersion': version,
+    'sonar.scanner.bootstrapStartTime': startTimestampMs.toString(),
+    // Bootstrap cache hit/miss is set later after the bootstrapper has run and before scanner engine is started
+    'sonar.scanner.wasJreCacheHit': 'false',
+    'sonar.scanner.wasEngineCacheHit': 'false',
+  };
+}
+
+/**
+ * Get endpoint properties from scanner properties.
+ */
+export function getHostProperties(properties: ScannerProperties): ScannerProperties {
+  let sonarHostUrl = properties[ScannerProperty.SonarHostUrl] ?? '';
+
+  if (!sonarHostUrl || SONARCLOUD_URL_REGEX.exec(sonarHostUrl)) {
+    return {
+      [ScannerProperty.SonarScannerInternalIsSonarCloud]: 'true',
+      [ScannerProperty.SonarHostUrl]:
+        properties[ScannerProperty.SonarScannerSonarCloudURL] ?? SONARCLOUD_URL,
+    };
+  }
+  return {
+    [ScannerProperty.SonarScannerInternalIsSonarCloud]: 'false',
+    [ScannerProperty.SonarHostUrl]: sonarHostUrl,
+  };
+}
+
+export function getProperties(
+  scanOptions: ScanOptions,
+  startTimestampMs: number,
+  cliArgs?: string[],
+): ScannerProperties {
+  const bootstrapperProperties = getBootstrapperProperties(startTimestampMs);
+  const cliProperties = getCommandLineProperties(cliArgs);
+  const scanOptionsProperties = getScanOptionsProperties(scanOptions);
+  const envProperties = getEnvironmentProperties();
+
+  // Compute default base dir respecting order of precedence we use for the final merge
+  const projectBaseDir =
+    cliProperties[ScannerProperty.SonarProjectBaseDir] ??
+    scanOptionsProperties[ScannerProperty.SonarProjectBaseDir] ??
+    envProperties[ScannerProperty.SonarProjectBaseDir] ??
+    process.cwd();
+
+  let inferredProperties: ScannerProperties;
+  try {
+    inferredProperties = getSonarFileProperties(projectBaseDir);
+  } catch (error) {
+    inferredProperties = {
+      'sonar.projectDescription': 'No description.',
+      'sonar.sources': '.',
+    };
+
+    const baseSonarExclusions =
+      cliProperties[ScannerProperty.SonarExclusions] ??
+      scanOptionsProperties[ScannerProperty.SonarExclusions] ??
+      envProperties[ScannerProperty.SonarExclusions] ??
+      DEFAULT_SONAR_EXCLUSIONS;
+
+    inferredProperties = {
+      ...inferredProperties,
+      ...getPackageJsonProperties(projectBaseDir, baseSonarExclusions),
+    };
+  }
+
+  // Merge properties respecting order of precedence
+  const properties = [
+    { 'sonar.projectBaseDir': projectBaseDir }, // Manually computed, can't be overridden
+    bootstrapperProperties, // Can't be overridden
+    cliProperties, // Highest precedence
+    scanOptionsProperties,
+    inferredProperties,
+    envProperties, // Lowest precedence
+  ]
+    .reverse()
+    .reduce((acc, curr) => ({ ...acc, ...curr }), {});
```

Comment:
why not just destructure them directly?

```suggestion
  const properties = {
    ...envProperties, // Lowest precedence
    ...inferredProperties,
    ...scanOptionsProperties,
    ...cliProperties, // Highest precedence
    ...bootstrapperProperties, // Can't be overridden
    ...{ 'sonar.projectBaseDir': projectBaseDir }, // Manually computed, can't be overridden
  }
```

Reply:
So, it's probably very questionable, but I wanted them written in the same order of priority they have, because otherwise my brain had issues writting this code in reverse order 😁

Now that it's written, we probably can indeed simplify this code to your suggestion.